### ISSUE TITLE
Allow \<newline> to escape the newline away

### ISSUE THREAD

**Issue by golddranks** [rust-lang-nursery/regex#107]
It's often desirable to use raw strings when declaring regular expressions. When raw strings are used, the standard Rust escapes are not in effect.

The regex syntax defines its own escapes, which are especially catered for escaping the regex puncuation characters, and some other characters that are difficult to include in the source code as-they-are. Raw strings are useful, because we don't have to double-escape the regex escapes, and can express characters like " freely.

However, because we can't use Rust escapes, we can't escape newlines either, like we can with normal strings:

> Line-break characters are allowed in string literals. Normally they represent themselves (i.e. no translation), but as a special exception, when a U+005C character \ occurs immediately before the newline, the U+005C character, the newline, and all whitespace at the beginning of the next line are ignored. Thus a and b are equal:

As far as I know, the regex crate doesn't have syntax that'd allow splitting the expression to multiple lines with escapable newlines: if there are newlines, they stand for themselves.

It would be nice to have an escape that allows that; backslash followed by newline character. It would help splitting raw string regexes to multiple lines for clarity. It would also help to follow coding conventions with regards to line length.

What do you think?

**Comment by golddranks** [comment]
Btw. currently backslash followed by newline just throws a parsing error, so I think that implementing that escape would be non-breaking change.

**Comment by BurntSushi** [comment]
Is there a reason why the `x` flag isn't suitable? See an example here: http://doc.rust-lang.org/regex/regex/index.html#example:-replacement-with-named-capture-groups and documented here: http://doc.rust-lang.org/regex/regex/index.html#grouping-and-flags

**Comment by golddranks** [comment]
Ah, I see, my premise that there isn't a way to ignore newline was wrong. Thanks for the info!

The x flag seems to provide the functionality I was missing - and more.

I'd still argue that supporting the backslash newline escape would be good, because it's simple and intuitive to use (in the sense that Rust strings, and many other languages support the same feature). I'd consider it as an additional ergonomic bonus - it's immediately clear what it does, and it's syntactically lightweight.

**Comment by BurntSushi** [comment]
That's not a bad point. I don't feel too strongly about it either way.

**Comment by golddranks** [comment]
I made a pull request that implements the feature. No pressure to accept it if there's some controversy about it, or if it needs further thought, wider acceptance, review etc.

**Comment by golddranks** [comment]
Quite a lot of time has passed. Any opinions on the \<newline> enhancement? Others than @BurntSushi ? Is there other heavy users and/or contributors that could provide their input?

**Comment by golddranks** [comment]
I can see that at least @Manishearth and @alexcrichton and @flying-sheep have frequently contributed here. What are your views on this?

**Comment by Manishearth** [comment]
(I only basically contribute rustups, I don't have strong opinions about the crate itself or understand how it works)

**Comment by golddranks** [comment]
The implementation was trivial, it's just that if we want to have one-to-one correspondence with RE2's syntax, this change is against that goal because RE2 doesn't implement this escape. However, my argument is that supporting \ newline escaping is something that 1) is nice and ergonomic 2) average user would expect to work 3) doesn't break anything and doesn't cause any further ramifications.

**Comment by BurntSushi** [comment]
I don't think a precise correspondence with RE2 syntax is a goal. I do think it is wise to be conservative and examine what other regex engines do. Do other regex engines support the type of newline escape being proposed here?

My concern is that we already have a solution to this problem using a technique established by several other regex engines (the `x` flag). Why add another?

**Comment by golddranks** [comment]
My motivation is to avoid monstrosity like this – to allow splitting the thing nicely into easy-to-understand parts without having to use the bulky (?x)(?-x) markup at the beginning and start of every line.

```
let dir_link = Regex::new(r##"<tr>(?x)
(?-x)<td colspan="2"><a href="/dirlist\.php\?dir=subtitles%2Fjapanese%2F([^"]+)%2F" class="">(?x)
(?-x)<strong>([^"]+)</strong> (?x)
(?-x)</a></td> (?x)
(?-x)<td class="tdright" title="([^"]+)" >[^<]+</td>(?x)
(?-x)</tr>"##)
```

I checked some other regex engines: JavaScript, C#, Python, Perl and Go all permit \<newline> escape but interpret that standing for just itself, i.e. newline. Some of the implementations (C#) document explicitly that any backslash escapes that are not in the spec, represent themselves.

**Comment by BurntSushi** [comment]
Some other thoughts:

1. I believe the current PR (#108) supports escaped new lines by simply dropping them. Isn't this at odds with other implementations, where escaped new lines are treated as themselves?
2. Rust has support for line continuation in normal strings, and that seems like the thing that is desired here. If we add something similar to regexes, shouldn't it function like Rust's existing line continuation?

These goals do seem at odds with each other.

Another related issue is #93, which suggests we should permit more escape sequences.

**Comment by golddranks** [comment]
Yes, they are at odds. The common theme in regex implementations seems to be that \<newline> represents <newline>. In the normal Rust strings \<newline> represents a null substring, it's simply dropped. This is for ergonomics, to allow long lines to be embedded in the source code. Of course, in raw strings, \<newline> always represents \<newline>, so the string is read as-is.

I was thinking in the lines of the following reasoning:

1) If in the regex crate, \<newline> drops the newline, that's different from how most other implementations do. That's a slight inconsistency.
2) However, is there any reason to have \<newline> to represent <newline>? I don't think so. It seems to be simply the accepted default. I can't think any situation where it would be sensible to use \<newline> to represent <newline>. If people want to write newline, they just write newline itself, or use the escape specially meant to represent newline.
3) Instead, there seems to be clear reason to have \<newline> to represent a null substring; to be able to end lines without affecting the string contents. Raw strings are often used with regexes, so it's good for ergonomics to implement this. Because \<newline> for <newline> doesn't seem to stand for any useful purpose, it would be more valuable if used for this.

So yes, I think it should behave like Rust's existing line continuation. (Does the current PR have some inconsistencies with that? If it does, I'm happy to fix it.)

Another way to think about it, is that when using raw strings, Rust basically delegates all decisions of escape sequences and such to the library that processes the string. If we want to give the users a "rusty" and ergonomic experience, we've got to re-implement ourselves some of the features Rust strings have. Of course, only those that make sense in the context of regexes.

**Comment by BurntSushi** [comment]
> However, is there any reason to have <newline> to represent ? I don't think so. It seems to be simply the accepted default. I can't think any situation where it would be sensible to use <newline> to represent . If people want to write newline, they just write newline itself, or use the escape specially meant to represent newline.

I feel like you're papering over (1). You're advocating for changing what is today a loud failure mode to a silent failure mode if someone is trying to port a regex over from another engine.

>  Instead, there seems to be clear reason to have <newline> to represent a null substring; to be able to end lines without affecting the string contents.

I don't think that's a clear reason at all. In particular, the `x` flag already allows one to do this. You have of course pointed out cases where the `x` flag isn't quite ideal, for example, when one wants to match on significant whitespace. This makes the need for newline escapes a bit more niche IMO.

> So yes, I think it should behave like Rust's existing line continuation. (Does the current PR have some inconsistencies with that? If it does, I'm happy to fix it.)

Indeed. Rust's existing line continuation drops the `\n` *and all proceeding whitespace up to the first non-whitespace character*.

To sum up, I think my primary concern here is that dropping escaped new lines will create a silent divergence from other regex engines. If we diverge, I'd rather it be in the form of an error returned to the caller. Alternatively, we could diverge if we thought it was important to do so. I'm not quite entirely convinced it's worth it in this case.

**Comment by golddranks** [comment]
Ah, I see. That's indeed a valid concern.

To me, the possibility that regexes that would be ported over to rust contain `\<newline>` seems very small, since it doesn't actually do anything worthwhile. (And fails to compile in some, like in RE2.)

Also, I have encountered quite a many large (not necessarily complicated) regexes that would greatly benefit from the simple way to "reformat" them a bit. (The `x` flag does that indeed, but it's visually heavier, eating away the niceness gained.) That's why I keep thinking this would be a worthwhile feature.

All in all, to me, the cons seem to be quite small, and pros large enough to outweigh them.

P.S. Thanks for pointing out the inconsistency with Rust's strings. That brings to mind: which would be more desirable kind of functionality for regexes, just dropping the newline, or doing it the Rust way? And again, then there would be a question about consistency with Rust's conventions. But that's a minor point compared to the main one.

**Comment by BurntSushi** [comment]
I like consistency better. Without it, you end up having to dedent your regex.
